pkg/kubestatemetrics/store/store.go

```diff
@@ -0,0 +1,202 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build kubeapiserver
+
+package store
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kube-state-metrics/pkg/metric"
+)
+
+// MetricsStore implements the k8s.io/client-go/tools/cache.Store
+// interface. Instead of storing entire Kubernetes objects, it stores metrics
+// generated based on those objects.
+type MetricsStore struct {
+	mutex sync.RWMutex
+	// metrics is a map indexed by Kubernetes object id, containing a slice of
+	// metric families, containing a slice of metrics.
+	metrics map[types.UID][]DDMetricsFam
+	// generateMetricsFunc generates metrics based on a given Kubernetes object
+	// and returns them grouped by metric family.
+	generateMetricsFunc func(interface{}) []metric.FamilyInterface
+
+	MetricsType string
+}
+
+// DDMetric represents the data we care about for a context.
+type DDMetric struct {
+	Labels map[string]string
+	Val    float64
+}
+
+// DDMetricsFam is the representation of a metric family.
+type DDMetricsFam struct {
+	Type        string
+	Name        string
+	ListMetrics []DDMetric
+}
+
+// NewMetricsStore returns a new MetricsStore.
+func NewMetricsStore(generateFunc func(interface{}) []metric.FamilyInterface, mt string) *MetricsStore {
+	return &MetricsStore{
+		MetricsType:         mt,
+		generateMetricsFunc: generateFunc,
+		metrics:             map[types.UID][]DDMetricsFam{},
+	}
+}
+
+func (d *DDMetricsFam) extract(f metric.Family) {
+	// f.Type is not extracted (value of gauge, counter etc) as we only support gauges.
+	d.Name = f.Name
+	for _, m := range f.Metrics {
+		var err error
+		s := DDMetric{}
+		s.Val = m.Value
+		s.Labels, err = buildTags(m)
+		if err != nil {
+			// TODO test how verbose that could be.
+			log.Errorf("Could not retrieve the labels for %s: %v", f.Name, err)
+			continue
+		}
+		d.ListMetrics = append(d.ListMetrics, s)
+	}
+}
+
+// Implementing k8s.io/client-go/tools/cache.Store interface
+// Add inserts adds to the MetricsStore by calling the metrics generator functions and
+// adding the generated metrics to the metrics map that underlies the MetricStore.
+func (s *MetricsStore) Add(obj interface{}) error {
+	o, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
```

Comment:
I think we can move the Lock() Unlock() near the line 98 since the function update the "metrics" map only in this ligne